File: 2020/C#/Day15.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AoC
{

    public static class Day15
    {
        public static void Part2(string[] lines)
        {
            int[] init = { 12, 20, 0, 6, 1, 17, 7 };

            Dictionary<int, int> nums = new Dictionary<int, int>();

            for (int i = 0; i < init.Length - 1; i++)
            {
                nums[init[i]] = i;
            }

            var lastNum = init.Last();

            for (int i = init.Length - 1; i < 30000000 - 1; i++)
            {
                int newNum;

                if (nums.TryGetValue(lastNum, out newNum))
                    newNum = i - newNum;
                else
                    newNum = 0;

                nums[lastNum] = i;

                lastNum = newNum;
            }

            Console.WriteLine($"{lastNum}");
        }
    }
}

File: 2020/day15.js
// var input = "12,20,0,6,1,17,7".split(",")
// var input = "0,3,6".split(",")

// var nums = []

// for (var i = 0; i < input.length; i++)
// {
	// nums.push(Number(input[i]))
// }

// var lastNum = nums.pop();

// for (var i = nums.length; i < 30000000; i++)
// {
	// var index = nums.lastIndexOf(lastNum)
	// if (index == -1)
		// index = nums.length;
	
	// nums.push(lastNum)
	
	// lastNum = nums.length - 1 - index;
// }

// console.log(nums[30000000 - 1])

var input = "12,20,0,6,1,17,7".split(",")
// var input = "0,3,6".split(",")

var nums = []

for (var i = 0; i < input.length - 1; i++)
{
	nums[Number(input[i])] = i
}

var lastNum = Number(input[input.length - 1])

for (var i = input.length - 1; i < 30000000 - 1; i++)
{
	var newNum = i - nums[lastNum];
	
	nums[lastNum] = i;
	
	if (isNaN(newNum))
		newNum = 0;
	
	lastNum = newNum;
}

console.log(lastNum)